runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Thu, Aug 27, 2015 at 10:35:58PM -0700, Andrew Vagin wrote:
>  lr-x------ 1 root root 64 Aug 27 16:14 0 -> /dev/null
>  lr-x------ 1 root root 64 Aug 27 16:14 2 -> /dev/null
>  l-wx------ 1 root root 64 Aug 27 16:14 255 -> pipe:[66755]
>
> How can we understand that 255 is external file descriptor…

Because it's an unnamed pipe associated with a process outside of the
container (see my ‘lsof -p {pid} +E’ for how to find the other
processes associated with the pipe [1]).

> … which is associated with the initial stdout?

For the general case (what is that thing on the other end?), I think
the semantic meaning is outside our specification [2].

For this specific case, #113 says that we'll pass through file
descriptors 0, 1, 2, and possibly more when we launch the application
process.  So we can find the semantics relative to the runtime by
inspecting the runtime or other hints.  Here's BusyBox running with
process.terminal set to true:

  # pstree -sSp 3958
  init(1)───login(2172)───…───sudo(3950)───runc(3954)───busybox(3958,ipc,mnt,pid,user,uts)

  # lsof -p 3958 +E
  COMMAND  PID  USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
  busybox 3958 wking  cwd    DIR   0,19       56 10149051 /
  busybox 3958 wking  rtd    DIR   0,19       56 10149051 /
  busybox 3958 wking  txt    REG   0,19  1947624 10149332 /bin/busybox
  busybox 3958 wking  mem    REG   0,18          10149332 /bin/busybox (path dev=0,19, inode=3013809)
  busybox 3958 wking    0u   CHR  136,6      0t0        9 /6
  busybox 3958 wking    1u   CHR  136,6      0t0        9 /6
  busybox 3958 wking    2u   CHR  136,6      0t0        9 /6
  busybox 3958 wking   10u   CHR    5,0      0t0       33 /dev/tty

From the DEVICE column, we can see that its standard streams are wired
up to /dev/pts/6.

Here's BusyBox running with process.terminal set to false:

  $ sudo lsof -p 4196 +E
  COMMAND  PID  USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
  runc    4192  root    3r  FIFO   0,10      0t0    18132 pipe 4192,runc,4w 4196,busybox,0r
  runc    4192  root    4w  FIFO   0,10      0t0    18132 pipe 4192,runc,3r 4196,busybox,0r
  runc    4192  root    5r  FIFO   0,10      0t0    18133 pipe 4192,runc,6w 4196,busybox,1w
  runc    4192  root    6w  FIFO   0,10      0t0    18133 pipe 4192,runc,5r 4196,busybox,1w
  runc    4192  root    7r  FIFO   0,10      0t0    18134 pipe 4192,runc,8w 4196,busybox,2w
  runc    4192  root    8w  FIFO   0,10      0t0    18134 pipe 4192,runc,7r 4196,busybox,2w
  busybox 4196 wking  cwd    DIR   0,19       56 10149051 /
  busybox 4196 wking  rtd    DIR   0,19       56 10149051 /
  busybox 4196 wking  txt    REG   0,19  1947624 10149332 /bin/busybox
  busybox 4196 wking  mem    REG   0,18          10149332 /bin/busybox (path dev=0,19, inode=3013809)
  busybox 4196 wking    0r  FIFO   0,10      0t0    18132 pipe 4192,runc,3r 4192,runc,4w
  busybox 4196 wking    1w  FIFO   0,10      0t0    18133 pipe 4192,runc,5r 4192,runc,6w
  busybox 4196 wking    2w  FIFO   0,10      0t0    18134 pipe 4192,runc,7r 4192,runc,8w

So you can see that its standard streams are wired up to runc's 3/4,
5/6, and 7/8, and that runC (as of v0.0.3) needs to do a better job
opening one-directional streams for this purpose ;).

[1]: https://github.com/opencontainers/specs/pull/87/files#r38044364
[2]: https://github.com/opencontainers/specs/pull/113#issuecomment-135639909